go-sdk/internal/sparse-merkle-tree/smt/merkletree.go

```diff
@@ -78,18 +80,33 @@ func (mt *sparseMerkleTree) AddLeaf(node core.Node) error {
 	// use up all the bits in the index's path. As soon as a unique path is found,
 	// which may be only the first few bits of the index, the new leaf node is added.
 	// One or more branch nodes may be created to accommodate the new leaf node.
-	newRootKey, err := mt.addLeaf(node, mt.rootKey, 0, path)
+	batch, err := mt.db.BeginBatch()
 	if err != nil {
 		return err
 	}
+	newRootKey, err := mt.addLeaf(batch, node, mt.rootKey, 0, path)
+	if err != nil {
+		log.L().Errorf("Error adding leaf node %s: %v, rolling back", node.Ref().Hex(), err)
+		_ = batch.Rollback()
+		return err
+	}
 	mt.rootKey = newRootKey
 
 	// update the root node index in the storage
-	err = mt.db.UpsertRootNodeIndex(mt.rootKey)
+	log.L().Infof("Upserting root node index to %s", mt.rootKey.Hex())
+	err = batch.UpsertRootNodeIndex(mt.rootKey)
 	if err != nil {
+		log.L().Errorf("Error upserting root node %s: %v, rolling back", mt.rootKey.Hex(), err)
+		_ = batch.Rollback()
+		return err
+	}
+	log.L().Infof("Committing batch operations for adding leaf node %s", node.Ref().Hex())
+	err = batch.Commit()
+	if err != nil {
+		log.L().Errorf("Error committing batch operations for adding leaf node %s: %v", node.Ref().Hex(), err)
+		_ = batch.Rollback()
```

Comment:
This is the core logic implements `Make merkle tree leaf addition operations atomic`. Some readability suggestions. Here are the steps as per my understanding:
1. A single DB tx is created to ensure atomicity of multiple DB txs 
2. Add a new leaf node as part of the started DB TX
3. Upsert MT root in the same DB TX
4. Commit only if both queries succeeded, otherwise rollback.

So I think it would improve the readability if `batch` is renamed to reflect it's a DB tx.

Reply:
Also, we should log Rollback errors as warnings so leakage of DB txs can be monitored.